File: backend/modules/production_saas_upgrade/refusal_engine.py
from __future__ import annotations

from typing import Any

from .document_generator import DocumentGenerator


class RefusalEngine:
    """Additive refusal engine for production SaaS module layer."""

    @staticmethod
    def execute(*, payload: dict[str, Any], actor_user_id: str, signature_method: str, signature_record: str, ip_address: str | None, device_info: str | None):
        template = "financial_liability_notice" if payload.get("document_kind") == "financial_notice" else "refusal_of_discharge"
        return DocumentGenerator.generate(
            template_key=template,  # type: ignore[arg-type]
            title="Discharge Refusal Legal Document",
            payload=payload,
            actor_user_id=actor_user_id,
            signature_method=signature_method,
            signature_record=signature_record,
            ip_address=ip_address,
            device_info=device_info,
        )

File: frontend/src/lib/server/platform/refusal-engine.ts
import type { Prisma } from "@prisma/client";
import { generateLegalDocument } from "@/lib/server/platform/document-automation";
import type { SignatureMethod } from "@/lib/server/platform/types";

export async function runRefusalEngine(input: {
  tenantId: string;
  actorUserId: string;
  caseId?: string;
  payload: Prisma.InputJsonValue;
  signatureMethod: SignatureMethod;
  signatureRecord: string;
  requestIp: string | null;
  deviceInfo: string | null;
}) {
  const templateKey = (input.payload as Record<string, unknown> | null)?.document_kind === "financial_notice"
    ? "financial_liability_notice"
    : "refusal_of_discharge";

  return generateLegalDocument({
    tenantId: input.tenantId,
    actorUserId: input.actorUserId,
    caseId: input.caseId,
    payload: input.payload,
    templateKey,
    title: "Discharge Refusal Legal Document",
    titleAr: "وثيقة الرفض الطبي القانونية",
    signatureMethod: input.signatureMethod,
    signatureRecord: input.signatureRecord,
    requestIp: input.requestIp,
    deviceInfo: input.deviceInfo,
  });
}
